src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+            string.push(c);
+            string.push_str(&n);
+            match &string[..] {
+                "nil" => Ok(Edn::Nil),
+                _ => Err(Error::ParseEdn(format!("{} cound not be parsed", string))),
+            }
+        }
+        _ => Err(Error::ParseEdn(
+            "Nullable boolean cound not be parsed".to_string(),
+        )),
+    }
+}
+
+fn read_vec(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(']') => return Ok(Edn::Vector(Vector::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_list(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(')') => return Ok(Edn::List(List::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_set(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    use std::collections::BTreeSet;
+    let mut res: BTreeSet<Edn> = BTreeSet::new();
+    loop {
+        match chars.next() {
+            Some('}') => return Ok(Edn::Set(Set::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.insert(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_map(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    use std::collections::BTreeMap;
+    let mut res: BTreeMap<String, Edn> = BTreeMap::new();
+    let mut key: Option<Edn> = None;
+    let mut val: Option<Edn> = None;
+    loop {
+        match chars.next() {
+            Some('}') => return Ok(Edn::Map(Map::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                if key.is_some() {
+                    val = Some(parse(Some(c), chars)?);
+                } else {
+                    key = Some(parse(Some(c), chars)?);
+                }
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+
+        if key.is_some() && val.is_some() {
+            res.insert(key.unwrap().to_string(), val.unwrap());
+            key = None;
+            val = None;
+        }
+    }
+}
+
+use std::borrow::Cow;
+
+pub trait MaybeReplaceExt<'a> {
```

Comment:
improves performance by `3 us`

Reply:
Is this trait necessary though? If you convert an `&str` to a `Cow` and do the default Rust `.replace` method, wouldn't it only change if the replace actually works?

I mean, the `Cow` stands for Copy On Write as far as I know, doesn't is support native checking for this? :thinking:

Reply:
it improved performance in 20%